### PR TITLE
update task manager docs after refactoring

### PR DESCRIPTION
**Pull request by fosterseth**
##### SUMMARY

issue https://github.com/ansible/awx/issues/13245

view rendered doc here https://github.com/fosterseth/awx/blob/update_task_manager_md/docs/task_manager_system.md

##### ISSUE TYPE

 - New or Enhanced Feature

##### COMPONENT NAME

 - API
 - Docs

### CODE REVIEW

File: docs/task_manager_system.md [hunk 1]
```diff
@@ -1,33 +1,113 @@
-# Task Manager Overview
+# Task Manager System Overview
 
-The task manager is responsible for deciding when jobs should be scheduled to run. When choosing a task to run, the considerations are:
+Made up of three separate components
```
Comment (kdelee) [inline docs/task_manager_system.md]:
```suggestion
The task management system is made up of three separate components:
```

File: docs/task_manager_system.md [hunk 2]
```diff
@@ -1,33 +1,113 @@
...
+
 
-## Task Manager Architecture
+### Task Manager Algorithm
+
+1. Get pending, waiting, and running tasks that have `dependencies_processed = True`
+2. For each running task, add task to dependency graph
+    a. dependency graph is just an internal data structure that tracks which
+    jobs are currently running. It also handles "soft" blocking logic
+3. Check if # tasks started on this task manager cycle is > `start_task_limit`
```
Comment (kdelee) [inline docs/task_manager_system.md]: missing also checking the task manager time out here

Comment (fosterseth) [reply docs/task_manager_system.md]: oh right, i'll add that in

File: docs/task_manager_system.md [hunk 3]
```diff
@@ -1,33 +1,113 @@
...
+Responsible for looking at each pending task and determining whether it can start that task.
+
 
-## Task Manager Architecture
+### Task Manager Algorithm
+
+1. Get pending, waiting, and running tasks that have `dependencies_processed = True`
+2. For each running task, add task to dependency graph
+    a. dependency graph is just an internal data structure that tracks which
+    jobs are currently running. It also handles "soft" blocking logic
```
Comment (kdelee) [inline docs/task_manager_system.md]: in same loop that we add tasks to dependency graph, we do the initial calculation of currently consumed capacity

```
 Before processing pending tasks, the task manager first processes running tasks. This allows it to build a dependency graph and account for the currently consumed capacity in the system.
    a. dependency graph is just an internal data structure that tracks which
    jobs are currently running. It also handles "soft" blocking logic
    b. the capacity is tracked in memory on the TaskManagerInstances and
        TaskManagerInstanceGroups objects which are  in-memory representations of the instances
        and instance groups. These data structures are used to help track what consumed capacity will
        be as we decide that we will start new tasks, and until such time that we actually commit
        the state changes to the database.
```

**Review by kdelee — COMMENTED** [review]
left a few suggestions

File: docs/task_manager_system.md (hunk 2 above)
Comment (fosterseth) [reply docs/task_manager_system.md]: updated the algorithm and added a `Timing Out` section

File: docs/task_manager_system.md (hunk 3 above)
Comment (fosterseth) [reply docs/task_manager_system.md]: added, thanks

File: docs/task_manager_system.md [hunk 4]
```diff
@@ -1,38 +1,127 @@
...
+                             |
+                             |
+                      +------v------+
+                      |   Project   |  dependency of Inventory Source Update B
+                      |   Update C  |  created by Dependency Manager
+                      +-------------+
+```
+
+
+### Dependency Manager Algorithm
```
Comment (gundalow) [inline docs/task_manager_system.md]: Is `Algorithm` the right term here, or is this a list of steps that are run through in order?

File: docs/task_manager_system.md [hunk 5]
```diff
@@ -1,38 +1,127 @@
...
 
 
 ### Job Lifecycle
 
 | Job Status |                                                       State                                                      |
-|:----------:|:------------------------------------------------------------------------------------------------------------------:|
+|:-----------|:-------------------------------------------------------------------------------------------------------------------|
 | pending    | Job has been launched.  <br>1. Hasn't yet been seen by the scheduler <br>2. Is blocked by another task <br>3. Not enough capacity |
 | waiting    | Job published to an AMQP queue.
```
Comment (AlanCoding) [inline docs/task_manager_system.md]: does pg_notify qualify as an AMQP queue? I don't think so?

File: docs/task_manager_system.md [hunk 6]
```diff
@@ -1,38 +1,127 @@
...
+4. Check and create dependencies for these newly created dependencies
+    a. inventory source updates can have a project update dependency
+5. All dependencies are linked to the parent task via the `dependent_jobs` field
+    a. This allows us to cancel the parent task if the dependency fails or is canceled
+6. Update the parent tasks with `dependencies_processed = True`
+
+
+## Task Manager
+
+Responsible for looking at each pending task and determining whether it can start that task.
```
Comment (gundalow) [inline docs/task_manager_system.md]: Shouldn't use `it` as it's ambiguous, do we mean:

```suggestion
Responsible for looking at each pending task and determining whether Task Manager can start that task.
```

File: docs/task_manager_system.md [hunk 7]
```diff
@@ -1,38 +1,127 @@
...
 
-## Task Manager Architecture
+### Task Manager Algorithm
```
Comment (gundalow) [inline docs/task_manager_system.md]: As above, if this is steps, then do we mean steps?

File: docs/task_manager_system.md [hunk 8]
```diff
@@ -1,38 +1,127 @@
...
+
+1. Get pending, waiting, and running tasks that have `dependencies_processed = True`
+2. Before processing pending tasks, the task manager first processes running tasks. This allows it to build a dependency graph and account for the currently consumed capacity in the system.
```
Comment (gundalow) [inline docs/task_manager_system.md:70]: Do all running tasks have to be processed before new pending tasks are processed?

Do we need to define "processing"?

File: docs/task_manager_system.md [hunk 9]
```diff
@@ -1,38 +1,127 @@
...
+    a. dependency graph is just an internal data structure that tracks which jobs are currently running. It also handles "soft" blocking logic
+    b. the capacity is tracked in memory on the `TaskManagerInstances` and `TaskManagerInstanceGroups` objects which are  in-memory representations of the instances and instance groups. These data structures are used to help track what consumed capacity will be as we decide that we will start new tasks, and until such time that we actually commit the state changes to the database.
+3. For each pending task:
+    a. Check if # tasks started on this task manager cycle is > `start_task_limit`
```
Comment (gundalow) [inline docs/task_manager_system.md]: Define `#` (if we mean number or ID, they say that)

File: docs/task_manager_system.md [hunk 10]
```diff
@@ -1,38 +1,127 @@
...
+2. Build up a workflow DAG for each workflow job
+3. For each workflow job:
+    a. Check if [timed out](#Timing Out)
+    b. Check if next node can start based on previous node status and the associated success / failure / always logic
+4. Create new task and signal start
+
+
+## Task Manager System Architecture
+
+Each of the three managers has a single entry point, `schedule()`. The method may be called in parallel, at any time, as many times as the user wants. The `schedule()` function tries to acquire a single, global lock using the Instance table first recorded in the database. If the lock cannot be acquired, the method returns. The failure to acquire the lock indicates that there is another instance currently running `schedule()`.
```
Comment (gundalow) [inline docs/task_manager_system.md]: "as many times as the user wants" what's the use-case for multiple parallel runs?

File: docs/task_manager_system.md [hunk 11]
```diff
@@ -46,19 +135,20 @@ Empirically, the periodic task manager has been effective in the past and will c
...
 
 
 ### Blocking Logic
 
 The blocking logic is handled by a mixture of ORM instance references and task manager local tracking data in the scheduler instance.
 
+There is a distinction between so-called "hard" vs "soft" blocking.
 
-## Acceptance Tests
+"Hard" refers to dependencies that are represented in the database via the task `dependent_jobs` field. That is, Job A will not run if any of its `dependent_jobs` are still running.
```
Comment (gundalow) [inline docs/task_manager_system.md]:
```suggestion
**Hard blocking** refers to dependencies that are represented in the database via the task `dependent_jobs` field. That is, Job A will not run if any of its `dependent_jobs` are still running.
```

**Review by murakobardenas — APPROVED** [review]

File: docs/task_manager_system.md [hunk 12]
```diff
@@ -46,19 +135,20 @@ Empirically, the periodic task manager has been effective in the past and will c
...
 
-The new task manager should, in essence, work like the old one. Old task manager features were identified while new ones were discovered in the process of creating the new task manager. Rules for the new task manager behavior are iterated below; testing should ensure that those rules are followed.
+"Soft" refers to blocking logic that doesn't have a database representation. Imagine Job A and B are both based on the same job template, and concurrent jobs is `disabled`. Job B will be blocked from running if Job A is already running. This is determined purely by the task manager tracking running jobs via the Dependency Graph.
```
Comment (gundalow) [inline docs/task_manager_system.md]:
```suggestion
**Soft blocking** refers to blocking logic that doesn't have a database representation. Imagine Job A and B are both based on the same job template, and concurrent jobs is `disabled`. Job B will be blocked from running if Job A is already running. This is determined purely by the task manager tracking running jobs via the Dependency Graph.
```


Can we reword "database representation"

**Review by murakobardenas — APPROVED** [review]

**Review by murakobardenas — APPROVED** [review]

File: docs/task_manager_system.md (hunk 10 above)
Comment (fosterseth) [reply docs/task_manager_system.md]: yeah this was inaccurate language. remove it and added a `bulk reschedule` section that addresses this topic

File: docs/task_manager_system.md (hunk 9 above)
Comment (fosterseth) [reply docs/task_manager_system.md]: fixed

File: docs/task_manager_system.md (hunk 7 above)
Comment (fosterseth) [reply docs/task_manager_system.md]: using `steps` now

File: docs/task_manager_system.md (hunk 5 above)
Comment (fosterseth) [reply docs/task_manager_system.md]: changed to
`Job submitted to dispatcher via pg_notify`

**Review by AlanCoding — APPROVED** [review]

**Pull request merged**